tools/preprocess_dataset_mpi.py

```diff
@@ -0,0 +1,572 @@
+                        builders[key].add_item(torch.IntTensor(sentence))
+                    builders[key].end_document()
+                    dset_stats[0] += 1
+                    dset_stats[1] += len(sentences)
+                dset_stats[2] += bytes_processed
+
+            if proc_rank == 0 and args.log_interval and count > 0 and count % args.log_interval == 0:
+                current = time.time()
+                elapsed = current - tokenize_start
+                mbs = dset_stats[2] / elapsed / 1024 / 1024 if elapsed > 0.0 else 0.0
+                docs = dset_stats[0]
+                print(f"Rank 0 processed {docs} documents",
+                      f"({docs/elapsed} docs/s, {mbs} MB/s).")
+                print(f"Estimated total processed {docs * proc_size} documents",
+                      f"({docs * proc_size / elapsed} docs/s, {mbs * proc_size} MB/s).",
+                      flush=True)
+
+        # finalize file of each rank
+        for key in columns:
+            builders[key].finalize(output_idx_files[key])
+            del builders[key] # file closed in __del__
+    except:
+        # caught an exception, assume our file is invalid
+        success = False
+
+    # wait for all ranks to finish their files
+    barrier(args)
+    tokenize_end = time.time()
+    all_sum_(args, dset_stats)
+    if proc_rank == 0:
+        secs = tokenize_end - tokenize_start
+        docrate = dset_stats[0] / secs if secs > 0.0 else 0.0
+        sentrate = dset_stats[1] / secs if secs > 0.0 else 0.0
+        byterate = dset_stats[2] / secs if secs > 0.0 else 0.0
+        print("Seconds to tokenize:", secs)
+        print("Documents=", dset_stats[0], "docs/sec=", docrate)
+        print("Sentences=", dset_stats[1], "sent/sec=", sentrate)
+        print("Bytes=", dset_stats[2], "bytes/sec=", byterate)
+
+    # allreduce to check whether all ranks wrote their part successfully
+    success = all_true(args, success)
+
+    # rank 0 merges and deletes all per-rank files
+    if proc_rank == 0:
+        # merge files if all ranks were successful
+        if success:
+            print("Merging rank files ...", flush=True)
+            merge_start = time.time()
+            numbytes = 0
+
+            # define name of single file
+            for key in columns:
+                filebase = "{}_{}_{}".format(args.output_prefix, key, level)
+                output_bin_files[key] = data_file_path(filebase)
+                output_idx_files[key] = index_file_path(filebase)
+                builders[key] = indexed_dataset.make_builder(output_bin_files[key],
+                                                             impl=args.dataset_impl,
+                                                             vocab_size=tokenizer.vocab_size)
+
+            # merge all ranks into one file
+            for rank in range(proc_size):
+                for key in columns:
+                    infile = "{}_{}_{}_{}".format(args.output_prefix, key, level, rank)
```

Comment:
Optional: I like f string better, don't know if you know about them:

```suggestion
                    infile = f"{args.output_prefix}_{key}_{level}_{rank}"
```

Personnal preference, feel free to ignore.

Reply:
I think the f string is easier to read, as well.